Docs and setup files

ascillitoe opened the pull request:
WIP, please ignore for now.

ascillitoe commented on alibi_detect/base.py:
```diff
@@ -122,6 +122,9 @@ def drift_config(self):
             cfg.update({'preprocess_fn': self.preprocess_fn})
 
+        # Add version_warning if it exists
+        cfg.update({'version_warning': self.meta.get('version_warning', False)})
```
If a `version_warning` was raised by `validate_config`, due to a `version` or `config_spec` mismatch, it is written to the config here. This ensures a warning will still be raised when the newly saved detector is reloaded again.

ascillitoe commented on alibi_detect/utils/schemas.py:
```diff
@@ -1,32 +1,30 @@
 #    SupportedModels.append()  # TODO
 SupportedModels = tuple(SupportedModels)
-SUPPORTED_MODELS = Union[SupportedModels]
+SupportedModels_py = Union[SupportedModels]  # only for use with pydantic. NOT to be used with mypy (as not static)
```
I've had to regress from the changes made in part II (see https://github.com/SeldonIO/alibi-detect/pull/462#discussion_r827852989). The part II solution didn't actually work properly when the `Union` was used in `saving/loading.py`. I suspect this is because we were trying to populate a `Union` with if-else statements, when mypy is a static type checker.

The new `SupportedModels_py` `Union` is constructed the same way, but is only used by pydantic. It seems using a dynamically generated `Union` is fine for that, but not for mypy. The type signatures used in `loading/saving.py` for models are relaxed to just be `Callable`....

ascillitoe commented on alibi_detect/utils/registry.py:
```diff
@@ -0,0 +1,24 @@
+
+# Register alibi-detect classes/functions
+if has_tensorflow:
+    registry.register('utils.tensorflow.kernels.GaussianRBF', func=GaussianRBF_tf)
```
Have kept population of the registry very simple for now. We'd likely want to readdress this if/when we wish to pre-register more functions/classes.

 On another related topic, I quite like the idea of adding thin wrappers for `GaussianRBF` and `preprocess_Drift`, which would select the correct backend-specific one. This means we could remove the burden on the user to pick the correct backend-specific option.

ascillitoe commented on alibi_detect/utils/fetching.py:
```diff
@@ -514,7 +514,7 @@ def fetch_detector(filepath: Union[str, os.PathLike],
             kwargs = get_pixelcnn_default_kwargs()
     detector = load_detector(filepath, **kwargs)
-    return detector
+    return detector  # type: ignore[return-value] # Since load_detector returns drift detectors but "Data" doesn't inc.
```
This ignore can be removed once all detectors use the config based approach.

jklaise replied:
Can you add this comment in the source code with a `TODO` please?

Also `but "Data" doesn't inc.`...?

ascillitoe commented on alibi_detect/utils/loading.py:
```diff
@@ -0,0 +1,625 @@
+    prep_model_and_emb as prep_model_and_emb_tf, load_kernel_config as load_kernel_config_tf, \
+    load_optimizer as load_optimizer_tf, load_embedding as load_embedding_tf, load_detector_legacy, Detectors
+#  from alibi_detect.utils.pytorch._loading import set_device
+import tensorflow as tf  # TODO - this is currently only required for FIELDS_TO_DTYPE conversion. Remove in future.
```
The `tensorflow` import will need to be removed in future in order to allow tensorflow to be an optional depedency.

ascillitoe commented on alibi_detect/utils/saving.py:
```diff
@@ -1,84 +1,27 @@
+from alibi_detect.utils.loading import load_detector as _load_detector  # TODO: Remove in future
+from alibi_detect.utils.tensorflow._saving import save_model_config as save_model_config_tf, save_detector_legacy
 import warnings
-import tensorflow as tf
```
All this moved to `utils/tensorflow/_saving.py`

ascillitoe commented on alibi_detect/utils/saving.py:
```diff
@@ -1,84 +1,27 @@
 
 logger = logging.getLogger(__name__)
 
-Data = Union[
```
All this moved to `utils/tensorflow/_saving.py`

ascillitoe commented on alibi_detect/utils/saving.py:
```diff
@@ -102,840 +47,18 @@ def save_detector(detector: Data, filepath: Union[str, os.PathLike]) -> None:
         logger.warning('Directory {} does not exist and is now created.'.format(filepath))
         filepath.mkdir(parents=True, exist_ok=True)
 
-    # save metadata
```
All this moved to `utils/tensorflow/_saving.py`

ascillitoe commented on alibi_detect/utils/tests/test_saving_legacy.py:
```diff
@@ -0,0 +1,246 @@
+"""
```
This is copied line for line from the original `test_saving.py`.

ascillitoe commented on alibi_detect/utils/tests/datasets.py:
```diff
@@ -0,0 +1,63 @@
+
+
+# Group dataset "cases" by type of data i.e. continuous, binary, categorical, mixed
+class ContinuousData:
```
Datasets are grouped together by data type with `pytest_cases`. We could adopt other strategies here, but this might be best left as a future discussion.

ascillitoe commented:
> * The current options for model type (e.g. `'UAE', 'HiddenOutput', 'custom'`) have been carried over from the legacy save/load code. We could do with rethinking what is required here.

The model `type` field wasn't that intuitive for use in the detector config (it was not clear why the type had to be `UAE` if an `embedding` was also spec'd, and the detector could not be saved when an `embedding` model was used with an additional model (as is done in the [amazon example](https://docs.seldon.io/projects/alibi-detect/en/latest/examples/cd_text_amazon.html)). Use of the `HiddenOutput` type was also not particuarly intuitive. 

The model `type` field has now been removed, and replaced with simpler logic: 
- When an `embedding` and `model` are spec'd, the model and embedding are automatically coupled in a `UAE` object, with the need for the user to specify this in the config. 
- Previously, spec'ing `type='HiddenOutput'` gave the same load behaviour as `type='custom'`. However, it would be desirable for `HiddenOutput` to accept a `layer` field, so that a full model could be spec'd in the config, along with the hidden layer to extract. This is now achieved with a new `layer` field - if `layer` is given, the spec'd model is wrapped in a `HiddenOutput`, and the requested layer is extracted.

#### Note
One side effect of the current `HiddenOutput` implementation is that if a `config.toml` is loaded with `layer` set, and the detector then saved again, the new `config.toml` will not contain the `layer` field. Instead, the `model` it references will be the extracted `HiddenOutput` model. It would be desirable to save the original `model` and `layer` field instead, as this would allow later configuration (i.e. change the extracted layer etc). However, this is not possible with the current `HiddenOutput` implementation since the original model is not stored.

ascillitoe commented:
> * Find an example where `custom_objects` is needed, and investigate how these objects are defined. It might be easier to remove support for this for now.

`custom_objects` is a dict to be passed to TensorFlow's [load_model](https://www.tensorflow.org/api_docs/python/tf/keras/models/load_model). Internally, we only currenty use it in one place:

https://github.com/SeldonIO/alibi-detect/blob/6b5b2c7b288e15039a115753d14720bb867c535a/alibi_detect/utils/fetching.py#L82-L86

but in general, it should be capable of passing custom objects/functions (e.g. for custom layers) to TensorFlow models. This is introduces another challenge; how do we save (and load) arbitrary custom objects stored inside `custom_objects`. We could support spec'ing the custom objects as runtime regsistered objects, as well as loading the custom objects from `.dill` (assuming they can be serialized as dill). 

#### Note
This still needs a little more thought. It is also going to be a point of difference between TensorFlow and PyTorch models.

jklaise commented on setup.py:
```diff
@@ -47,6 +47,8 @@ def readme():
         "requests>=2.21.0, <3.0.0",
         "numba>=0.50.0, !=0.54.0, <0.56.0",  # Avoid 0.54 due to: https://github.com/SeldonIO/alibi/issues/466
         "pydantic>=1.8.0, <2.0.0",
+        "toml>=0.10.1, <0.11.0",  # STC, see https://discuss.python.org/t/adopting-recommending-a-toml-parser/4068
+        "catalogue>=2.0.0, <2.0.7",
```
I don't think we need to be so strict here, why not `<1.0.0` and `<3.0.0`?

jklaise commented:
@ascillitoe on the submodules for `save_detector`, `load_detector`, I think it's about time we exported these into a higher public API rather than the user having to import from `alibi_detect.utils.saving/loading`? In `alibi` I've just done `alibi.saving`, if we wanted to do the same perhaps a new subpackage `alibi_detect.saving` could be defined that imports the public functions from the deeper modules?

jklaise commented:
@ascillitoe quite a few modules under `utils` (including but not limited to `saving.py` and `loading.py` are not rendering on the API docs.

jklaise commented on doc/source/conf.py:
```diff
@@ -108,6 +108,8 @@
     "transformers",
     "tqdm",
     "dill",
+    "pydantic",
```
After the docstrings are rendering, would make sure that mocking `pydantic` doesn't result in weird things, i.e. would be nice to have readable `pydantic` models in the API docs.

ascillitoe replied:
This is resolved implicitly since we are now using `sphinx-autoapi` (https://github.com/SeldonIO/alibi-detect/pull/482), meaning modules do not require mocking.

jklaise commented on doc/source/cd/methods/cvmdrift.ipynb:
```diff
@@ -94,16 +94,33 @@
    "cell_type": "markdown",
-   "metadata": {},
+   "metadata": {
+    "pycharm": {
+     "name": "#%% md\n"
```
Out of curiosity what is this metadata? I assume it doesn't matter if it's here in the published versions?

ascillitoe replied:
This is accidental. Think it was added as I opened the file in PyCharm. Will remove.

jklaise commented on doc/source/cd/methods/spotthediffdrift.ipynb:
````diff
@@ -189,6 +189,26 @@
+    "cd = load_detector(filepath)\n",
+    "```\n",
+    "\n",
+    "Currently on the **TensorFlow** backend is supported for `save_detector` and `load_detector`. Adding **PyTorch** support is a near term priority."
````
on -> only ?

jklaise commented on doc/source/overview/config_files.md:
```diff
@@ -0,0 +1,753 @@
+previously created detectors.
+- **Flexible artefact specification**: Artefacts such as datasets and models can be specified as locally serialized
+objects, or as runtime registered objects (see [Specifying complex fields](complex_fields)). Multiple detectors can 
+share the same artefacts, and they can be easily swapped.
```
Can we claim this already in practice? I feel like there would be some relative path and/or symlink shenanigans...

ascillitoe replied:
I had in mind that absolute file paths would be used in this case. I did originally implement to allow absolute paths but admittedly haven't tested recently. I will test again if we decide we want to retain this capability (otherwise will remove mentions in the docs).

ascillitoe replied:
I've added a test `test_load_absolute` to test that `load_detector` works with absolute paths in the `config.toml`.

jklaise commented:
@ascillitoe also thinking if we should elevate the registry from `alibi_detect.utils.registry` to e.g. `alibi_detect.registry`?

jklaise commented on doc/source/overview/config_files.md:
````diff
@@ -0,0 +1,753 @@
+|`initial_diffs`           |✔          |            |        |          |
+```
+
+```{note}
+When TensorFlow, PyTorch, or scikit-learn `model`'s are specified, the model type should be specified via the `backend` field in the *config.toml*.
+For example, if `model` is a TensorFlow model, set `backend='tensorflow'. This is the case even for detectors that don't take a `backend` kwarg.
+```
````
This doesn't render properly: 
![image](https://user-images.githubusercontent.com/13080878/160413211-b863f15a-273f-4662-8401-ea7b299f0ab3.png)

Also can you clarify the last sentence? Does it mean `backend='tensorflow'` should be put for everything by default?

jklaise replied:
Ok, let's clarify on the public docs and perhaps leave a note somewhere in the source code about this behaviour?

ascillitoe replied:
p.s. on a side note, this is related to the first "Outstanding Decisions" item. i.e. do we want to factor out `backend` into top-level detector one, `model` one, and `preprocess_fn` one. I don't think we need to right now, as mixing backends doesn't currently make sense. But it will become necessary when we introduce backends which are distinct from model types (e.g. `keops`). Things will also get more complicated here though, as for e.g. `keops` supports PyTorch but not TensorFlow (so the detector `backend` will restrict the options for the `preprocess_fn` "backend"). 

Given the complexity, I'm tempted to leave for now (and just note as you mentioned)?

jklaise commented on doc/source/overview/config_files.md:
```diff
@@ -0,0 +1,753 @@
+### Registering artefacts
+
+Custom artefacts defined in Python code may be specified in the config file without the need to serialise them, 
+by first adding them to the Alibi Detect artifact registry using the [registry](../api/alibi_detect.utils.registry.rst) 
```
artifact -> artefact (the only place with an `i` !)

jklaise commented on doc/source/overview/config_files.md:
```diff
@@ -0,0 +1,753 @@
+
+#### Models 
+
+A `model` dictionary may be used to specify a model within a `preprocess_fn`, or a model for detectors such as the
```
Are we happy with the semantics that a stand-alone `model` dictionary might in fact refer to one that is an argument to `preprocess_fn`? I'm not sure I'm 100% comfortable with this (and this also goes beyond our tacit assumption of keeping config files API-compatible), also what if there is both a `model` and a `preprocess_fn.model`? @arnaudvl

ascillitoe replied:
As it currently stands if there were a `model` this would be passed as a detector kwarg, and if there were also a `preprocess_fn.model` this would be passed as a kwarg to `preprocess_drift`, which would then be passed into the detector `preprocess_fn`. I've just modified the `test_save_classifierdrift` to test this (i.e. it saves and loads a detector with a `model` and `preprocess_fn.model`). 

I sort of think this is the most API-compatible way to do it? Since we do indeed have a kwarg named `model` in detectors and `preprocess_drift`? We could rename the field in `preprocess_fn`, but that wouldn't match the `preprocess_drift` API. 

On a related note, the general idea of the config functionality is that the artefact dictionaries are modular/portable i.e. the same `ModelConfig` can be used to define any model, regardless of whether it is passed as a detector kwarg, a `preprocess_fn` kwarg, or even to the `proj` kwarg of `DeepKernel`. The same applies for the `KernelConfig`, which might be a detector kwarg, or a kwarg of `DeepKernel`. I'm not sure right now I see a use case for making different model, kernel config's etc for different use cases? Although I could have missed some use cases...

jklaise replied:
Oh maybe I've been misled this whole time... My impression was that if I specify e.g. `[model]` dictionary at top-level, it would automagically be inferred as actually being `[preprocess_drift.model]`? I guess from what you've said now that is not actually the case?

ascillitoe replied:
p.s. just to be clear, if the above was a `ClassifierDrift` config which takes a `model` kwarg, we would have just added a field such as `model.src = "classifier_model/"`. (note that we can skip the [ ] notation and condense onto fewer lines when desired).

jklaise replied:
Hmm maybe, the key is to distinguish between a `model` (top-level) field  and a `model` dictionary maybe? Tho it's the same thing if `model` is top-level.

ascillitoe replied:
OK I'll have a think. Ideally I'd like to get rid of lots of this "artefact dictionaries" stuff from the docs page, and offload it to the api docs page. Then be more succinct about the modular "drop-in" type approach with these dicts such as `model`. However, I've hidden `schemas.py` from the API docs for now as it doesn't render very nicely due to all the inheritance in there...

ascillitoe replied:
This is resolved in the new docs.

jklaise commented on doc/source/overview/config_files.md:
```diff
@@ -0,0 +1,753 @@
+  
+* - layer
+  - Optional index of hidden layer to extract. If not `None`, a 
+  [HiddenOutput](../api/alibi_detect.cd.tensorflow.preprocess.rst) model is returned.
```
Maybe this is one for docstrings too, but I'm not sure what it means for a `HiddenOutput` model to be returned?

ascillitoe replied:
Thanks, will do.

jklaise commented on doc/source/overview/config_files.md:
```diff
@@ -0,0 +1,753 @@
+  [HiddenOutput](../api/alibi_detect.cd.tensorflow.preprocess.rst) model is returned.
+  - `None`
+  
+* - custom_obj
+  - Dictionary of custom objects. Passed to the tensorflow 
+  [load_model](https://www.tensorflow.org/api_docs/python/tf/keras/models/load_model) function. This can be used to
+  pass custom registered functions and classes to a model.
+  - `None`
```
Placeholder for further discussion.

ascillitoe commented on doc/source/_static/code_blocks.css:
```diff
@@ -0,0 +1,17 @@
+.codeblock-label{
+	display:inline-block;
+	border-top-left-radius:.5rem;
+	border-top-right-radius:.5rem;
+	font-family:Menlo,Monaco,Consolas,Liberation Mono,Courier New,monospace;
+	font-size:.75rem;--bg-opacity:1;
+	background-color:#e1e7ef;
+	background-color:rgba(224.70600000000002,231.07080000000002,239.394,var(--bg-opacity));
+	padding:.25rem .75rem;
+	border-top-width:1px;
+	border-left-width:1px;
+	border-right-width:1px;--border-opacity:1;
+	border-color:#ccd6e0;
+	border-color:rgba(204,213.85999999999999,224.39999999999998,var(--border-opacity));
+	margin-bottom:0;
+}
+
```
"Nice" is strongly subjective! This is actually only for the `config.toml` box which is added to toml file snippets. I wanted a way to differentiate them from standard code snippets as otherwise it was confusing.

![image](https://user-images.githubusercontent.com/32061685/160402349-c896a4ed-abd1-45de-a756-ab52747585a8.png)

Might tweak this at a later date but didn't want to overly complicate for now.

jklaise replied:
Ah I see, wonder if there's a way to leave comments in the `.css` file so we know the intended use case? Or potentially more descriptive class name?

jklaise commented on doc/source/overview/config_files.md:
```diff
@@ -0,0 +1,753 @@
+from alibi_detect.utils.loading import load_detector
+
+# Register a simple function
+@registry.register('my_function.v1')
```
Slightly off-topic, what happens if no name is passed to the function? I would think a nice default would be to automatically register with the function name, but not sure if `catalogue` does it. If not then no worries.

jklaise replied:
OK we can open an issue afterwards.

jklaise commented on alibi_detect/utils/registry.py:
```diff
@@ -0,0 +1,24 @@
+    from alibi_detect.cd.pytorch import preprocess_drift as preprocess_drift_torch
+
+# Create registry
+registry = catalogue.create("alibi_detect", "registry")
```
Let's move the module to `alibi_detect.registry` as discussed imo.

jklaise commented on alibi_detect/utils/validate.py:
```diff
@@ -34,6 +34,12 @@ def validate_config(cfg: dict, resolved: bool = False) -> dict:
         raise ValueError(f'Loading the {detector_name} detector from a config.toml is not yet supported.')
 
+    # Raise warning if config file already contains a version_warning
+    version_warning = cfg.get('version_warning', False)
```
Is it enough to just save a boolean flag or perhaps it would be wiser to save the original version of the library (and `config_spec`?) into an immutable config field?

ascillitoe replied:
I didn't think of that, it'll serve as nice record-keeping in the config too. Will add.

ascillitoe replied:
@jklaise @mauicv I started to rework this to save an immutable `orig_version` and `orig_config_spec` field instead of `version_warning`. But after thinking about it more, I'm having second thoughts.

For an example: imagine we save a config file with `version=0.9.1`. Later we load it in version `0.10.0`, this would raise a warning as desired. However, if we saved the config again, we'd rewrite `version=0.10.0`. Therefore on subsequent loads we wouldn't be warned that there is potentially a problem. This is where `version_warning` comes in; it would be written to the config upon the first version mismatch, and then will cause a warning to be raised on subsequent loads regardless of `version`. 

Onto the proposed `orig_version`, I'm wondering if this extra info would just be confusing. In the above scenario, we'd write `orig_version=0.9.1`. A warning would be correctly raised on first load with `version=0.10.0`. However, the problem scenario is if we then saved a config, and loaded it back in with version 0.9.1 installed. This would still raise a warning since `version=0.10.0` in the config but the current version is 0.9.1. However, I think might lead to a false sense of security (or just be confusing) now we have `orig_version=0.9.1` and current installed version 0.9.1. 

TL;DR - what would this extra `orig_version` info actually give us? It seems to me that it might just lead to confusion...

jklaise replied:
I agree with the sentiment. I think we might be missing a generally accepted pattern here as this problem would come up any time someone implements saving/loading functionality somewhere, perhaps it would be useful to briefly survey the problem.

Otherwise I agree that it might be less confusing to just stick with `version_warning`.

jklaise commented on alibi_detect/utils/schemas.py:
```diff
@@ -1,32 +1,32 @@
-# SupportedModels is a tuple of the real class types, to be used in isinstance() in saving.py and loading.py.
-# SUPPORTED_MODELS is a typing.Union of the same model types, to be used in type hints.
-SupportedModels = [UAE, HiddenOutput]
+# SupportedModels_types is a tuple of possible models (conditional on installed deps). This is used in isinstance() etc.
```
SupportedModels_types -> SupportedModels ?

jklaise commented:
@ascillitoe can I suggest running `isort` at least on the new modules? I know we don't have it in the official guidelines, but perhaps we should...

jklaise commented on alibi_detect/utils/loading.py:
```diff
@@ -0,0 +1,625 @@
+    ['initial_diffs']
+]
+
+# Fields to convert from list to tuple in resolve_cfg
+FIELDS_TO_TUPLE = [
+    ['detector', 'kwargs', 'input_shape']
+]
+
+# Fields to convert from list to np.ndarray in resolve_cfg
+FIELDS_TO_ARRAY = [
+    ['sigma'],
+    ['kernel', 'sigma'],
+    ['kernel', 'kernel_a', 'sigma'],
+    ['kernel', 'kernel_b', 'sigma'],
+]
+
+# Fields to convert from str to np.dtype
+FIELDS_TO_DTYPE = [
+    ['preprocess_fn', 'dtype']
+]
```
I have yet to read the rest of the logic so perhaps this comment is premature - but I thought `pydantic` should take care of this (e.g. feeding a `list` into a `pydantic` model that expects a `tuple` would coerce into a `tuple`; this can be extended to custom validators so we can also do `list` -> `np.ndarray` coercion).

ascillitoe replied:
That's a good point, I shall explore this.

ascillitoe replied:
I'm now using pydantic to coerce to `Tuple` and `np.ndarray` as suggested. I've left `FIELDS_TO_DTYPE` as is for now.

Note: I use the new `NDArray` validator when typing simple ndarray's in the pydantic models (such as for `sigma`). I've left `x_ref` typed with the less strict `np.ndarray` for now, as would need to think about how to extend the `NDArray` class to accept `Any` for this.

ascillitoe replied:
@jklaise leaving this unresolved so that you can check you are happy with the above.

jklaise replied:
Nice! Looks much cleaner. Left a couple of comments on the `NDArray` type validator too.

jklaise commented on alibi_detect/utils/loading.py:
```diff
@@ -0,0 +1,625 @@
+]
+
+
+def load_detector(filepath: Union[str, os.PathLike], **kwargs) -> Detectors:
```
Can we document the use cases for supplying `kwargs` here?

ascillitoe replied:
This has actually made me realise we (I) might need to think again about the `custom_objects` field. The `kwargs` is passed to `load_detector_legacy`, which accepted additional generic kwarg's for a number of things. The main one of interest is that if `custom_objects` was in `kwargs`, it would be passed to `load_model`:

https://github.com/SeldonIO/alibi-detect/blob/fc83c707d359d4e5009b7cfb84191748a5625bdd/alibi_detect/utils/tensorflow/_loading.py#L310-L311

So actually what I currently have where `custom_object` is read/written to `config.toml` perhaps doesn't make sense, as the existing functionality is actually to pass in as kwarg to `load_detector`. But we could support spec'ing it via config as a pre-registered function/class... but then I'm wondering if it makes more sense to explore how well the existing tensorflow custom object registering works here https://www.tensorflow.org/api_docs/python/tf/keras/utils/register_keras_serializable.

TL;DR I'm wondering if we should just remove `custom_object` support entirely for now, and add back in as a later addition?

ascillitoe replied:
`custom_objects` would only be used if it needed to be passed to the tensorflow load function. We currently only do this for our `resnet56` model (see https://github.com/SeldonIO/alibi-detect/pull/469#issuecomment-1080434048).

If we removed I can't see there being many issues in the short-term, except that we are technically reducing capability (as the legacy `load_detector` accepted `custom_objects` as a kwarg).

jklaise commented on alibi_detect/utils/loading.py:
```diff
@@ -0,0 +1,625 @@
+        elif 'meta.pickle' in files:
+            return load_detector_legacy(filepath, '.pickle', **kwargs)
+        else:
+            raise ValueError('Neither meta.dill, meta.pickle or config.toml exist in {}.'.format(filepath))
```
Let's use f-strings as default please :)

ascillitoe commented on doc/source/overview/config_files.md:
```diff
@@ -0,0 +1,753 @@
+|:---------------------------------------------------------------------|:----------------------------------------------|------------|---------|
+| [preprocess_drift](../api/alibi_detect.cd.tensorflow.preprocess.rst) | `'@cd.[backend].preprocess.preprocess_drift'` | ✔           | ✔       |
+| [GaussianRBF](../api/alibi_detect.utils.tensorflow.kernels.rst)      | `'@utils.[backend].kernels.GaussianRBF'`      | ✔           | ✔       |
+| [TFDataset](../api/alibi_detect.utils.tensorflow.data.rst)           | `'@utils.tensorflow.data.TFDataset'`          | ✔           |         |
```
Its used as a default kwarg in a number of detectors, e.g.

https://github.com/SeldonIO/alibi-detect/blob/6b5b2c7b288e15039a115753d14720bb867c535a/alibi_detect/cd/tensorflow/classifier.py#L37

ascillitoe commented on alibi_detect/utils/loading.py:
```diff
@@ -0,0 +1,625 @@
+        config_dir = config_file.parent
+        cfg = read_config(config_file)
+    else:
+        raise ValueError("`filepath` should point to a directory containing a 'config.toml' file.")
```
Yep that was my thinking!

jklaise commented on alibi_detect/utils/loading.py:
```diff
@@ -0,0 +1,625 @@
+
+    # Get x_ref
+    x_ref = cfg.pop('x_ref')
+    if isinstance(x_ref, str):  # If x_ref still a str, resolving must have failed
+        raise ValueError("Failed to resolve x_ref field.")
```
Under what circumstances is this possible? Shouldn't the error be raised earlier by the resolution code?

ascillitoe replied:
Removing this (and similar checks you noticed below), since we expect `resolve_config` to be passed a validated config. In other words, `validate_config(cfg, resolved=True)` will already have raised a `ValidationError` if `x_ref` is not an `np.ndarray`.

jklaise commented on alibi_detect/utils/loading.py:
```diff
@@ -0,0 +1,625 @@
+    if model is not None:
+        if not isinstance(model, SupportedModels):
+            raise ValueError(f"Failed to load the {detector_name}'s model."
+                             "Is the `model` field specified, and is the model a supported type?")
```
On second thought, is it even possible that a resolved model can be of unsupported type...?

ascillitoe replied:
Removed. See https://github.com/SeldonIO/alibi-detect/pull/469#discussion_r849577745.

jklaise commented on alibi_detect/utils/loading.py:
```diff
@@ -0,0 +1,625 @@
+    detector = _init_detector(x_ref, cfg, preprocess_fn=preprocess_fn, model=model, kernel=kernel, backend=backend)
+
+    # Update metadata
+    detector.meta.update({'config_file': str(config_file.resolve())})
```
It seems to me we're not adding this metadata for normally (non-config) instantiated detectors? Also what is the use case of having a plain string in the metadata? Any manipulation would be done by the detector's `get_config` method.

ascillitoe replied:
This is simply adding the absolute filepath of the `config.toml` file to the metadata, so that the user could see where the instantiated detector came from at runtime. Can remove if you think unnecessary!

jklaise replied:
Not sure if needed as the user wouldn't be able to load a detector if they didn't know the path already? Also, as I mentioned, this introduces a disparity in the metadata between runtime- and config-defined detectors.

ascillitoe replied:
Good point, I'll remove!

ascillitoe replied:
p.s. @jklaise, on a related note, we need to add the `version_warning` field to the detector somewhere, so that we have access to it at save time. I currently add this in to `detector.meta`. Do you think this should go somewhere else, i.e. just as `detector.version_warning`? I get the impression you're thinking about us having consistent detector metadata across all use cases etc?

jklaise commented on alibi_detect/utils/loading.py:
```diff
@@ -0,0 +1,625 @@
+    layer = cfg['layer']
+    src = Path(src)
+    if not src.is_dir():
+        raise FileNotFoundError("Compatible model not found at %s" % str(src.resolve()))
```
Ah string interpolation as well, let's stick with f-strings please :)

jklaise commented on alibi_detect/utils/loading.py:
```diff
@@ -0,0 +1,625 @@
+    if backend == 'tensorflow':
+        model = load_model_tf(src, load_dir='.', custom_objects=custom_obj, layer=layer)
+    else:
+        raise NotImplementedError('Loading of pytorch models not currently supported')
```
Do we need to add `sklearn` here? Or are we not really doing that anyway?

jklaise commented on alibi_detect/utils/loading.py:
```diff
@@ -0,0 +1,625 @@
+    typ = cfg['type']
+    if backend == 'tensorflow':
+        emb = load_embedding_tf(src, embedding_type=typ, layers=layers)
+    else:
+        emb = None
```
Noting that the logic here is different than for loading the model, i.e. returning `None` instead of raising an error.

jklaise commented on alibi_detect/utils/loading.py:
```diff
@@ -0,0 +1,625 @@
+    filepath = Path(filepath)
+
+    cfg = dict(toml.load(filepath))  # toml.load types return as MutableMapping, force to dict
+    logger.info('Loaded config file from %s', str(filepath))
```
I knew that logging was an exception to f-strings due to performance overhead :(

ascillitoe replied:
I've tried to replace all string interpolations. We should either have f-strings for error raising and `.format()`'s for logging.

jklaise commented on alibi_detect/utils/loading.py:
```diff
@@ -0,0 +1,625 @@
+    cfg = dict(toml.load(filepath))  # toml.load types return as MutableMapping, force to dict
+    logger.info('Loaded config file from %s', str(filepath))
+
+    # Convert keys in categories_per_feature back to str. This is undesirable, but currently necessary in order
```
I think the code is doing the opposite - converting from `str` to `int`.

Again, using `pydantic` for validation, the conversion should happen automatically due to type coercion, I wonder why we need to do it here explicitly?

ascillitoe replied:
Thanks! Just tested and this is no longer needed (think it was left over from before I added pydantic validation).

jklaise commented on alibi_detect/utils/loading.py:
```diff
@@ -0,0 +1,625 @@
+    ----------
+    cfg
+        The unresolved config dict.
+    config_dir
+        Filepath to directory the `config.toml` is located in. Only required if different from the
+        runtime directory, and artefacts are specified with filepaths relative to the config.toml file.
```
What would be an example of using this functionality? Is this function intended for public usage as well?

ascillitoe replied:
Do you mean using `resolve_config` in general? Or the `config_dir` kwarg?

To answer the first, I've left `read_config`, `resolve_config` and `validate_config` public so that advanced users can peform the `load_detector` workflow manually for troubleshooting i.e. `cfg = read_config()` -> `cfg = validate_config(cfg)` -> *custom fetching* -> `cfg = resolve_config(cfg)` -> `cfg = validate_config(cfg, resolved=True)`.

The above could also be useful if *custom fetching* of artefacts is required. However, the missing piece of the puzzle here is that `load_detector` currently only accepts a file path to a ` config.toml` (I originally had it accepting resolved `cfg` dicts too, so it could complete the above workflow. This is what I hand in mind as the full workflow if `config.toml` files contained refs to remote artefacts). We could make `resolve_config` etc private for now and come back to this at a later date if needed? Or we could just leave as is... 

Wrt to `config_dir`, this is used by default internally, since the we pass `load_detector` a `filepath` containing the `config.toml`, which would usually be different to the runtime directory i.e. `filepath='my_detector'` instead of `filepath='.'`.

jklaise commented on alibi_detect/utils/loading.py:
```diff
@@ -0,0 +1,625 @@
+    return cfg
+
+
+def resolve_cfg(cfg: dict, config_dir: Optional[Path]) -> dict:
```
Let's try to keep consistent with names, e.g. currently `read_config` vs `resolve_cfg`.

jklaise commented on alibi_detect/utils/loading.py:
```diff
@@ -0,0 +1,625 @@
+                if src in registry.get_all():
+                    obj = registry.get(src)
+                else:
+                    raise ValueError("Can't find %s in the custom function registry" % src)
```
f-string!

jklaise commented on alibi_detect/utils/loading.py:
```diff
@@ -0,0 +1,625 @@
+    return cfg
+
+
+def resolve_cfg(cfg: dict, config_dir: Optional[Path]) -> dict:
```
I'm slightly concerned that this function is quite complex (even though well documented). Perhaps worth splitting out the logic in even more functions (maybe using `pydantic` type coercion would also reduce the body). Also should we unit test at least some of these utility functions?

ascillitoe replied:
I've stripped it down a bit by removing the original explicit type conversion bits, and factoring out the dtype conversion to a private function (we could also investigate using pydantic for the dtype conversion in the future). 

The function is still pretty complex though...

jklaise commented on alibi_detect/utils/tensorflow/_loading.py:
```diff
@@ -0,0 +1,1054 @@
+        # Assemble deep kernel
+        kernel = DeepKernel(proj, kernel_a=kernel_a, kernel_b=kernel_b, eps=eps)
+    else:
+        raise ValueError('Unable to process kernel.)')
```
Probably worth spelling out that the only two options are `src` and `proj`?

jklaise commented on alibi_detect/utils/saving.py:
```diff
@@ -88,12 +31,14 @@ def save_detector(detector: Data, filepath: Union[str, os.PathLike]) -> None:
-    if detector_name not in DEFAULT_DETECTORS:
+    detector_name = detector.__class__.__name__
+    if detector_name not in [detector.__name__ for detector in get_args(Detectors)]:
         raise ValueError('{} is not supported by `save_detector`.'.format(detector_name))
```
Change to `NotImplementedError` and f-string?

jklaise commented on alibi_detect/utils/tensorflow/_loading.py:
```diff
@@ -0,0 +1,1054 @@
+
+# This list is currently defined in here to avoid circular dependency (it is needed in utils/loading.py). Once the
+# legacy loading support is removed, this will be moved to utils/loading.py
+Detectors = Union[
```
nitpick: since this is a type should we use singular `Detector` ?

mauicv commented on alibi_detect/utils/loading.py:
```diff
@@ -0,0 +1,625 @@
+    layer = cfg['layer']
+    src = Path(src)
+    if not src.is_dir():
+        raise FileNotFoundError("Compatible model not found at %s" % str(src.resolve()))
```
Oh, the error message could be an f-string however!

mauicv replied:
I think these checks are actually done in `load_model_tf` below? But there they raise warnings instead? Why don't we raise the errors there?

ascillitoe replied:
Thanks. I've reworded the `FileNotFoundError` as recommended, and deleted the similar warning in `tensorflow.loading.load_model`.

jklaise commented on alibi_detect/utils/saving.py:
```diff
@@ -948,856 +71,384 @@ def load_detector(filepath: Union[str, os.PathLike], **kwargs) -> Data:
-    Parameters
-    ----------
+    # Save x_ref
+    save_path = filepath.joinpath('x_ref.npy')
+    np.save(str(save_path), cfg['x_ref'])
+    cfg.update({'x_ref': 'x_ref.npy'})
```
Minor comment: perhaps worth defining constants at the top of the module for repeatedly used strings, e.g. `X_REF_FILENAME='x_ref.npy'`

ascillitoe replied:
Done for `'x_ref.npy'`. I Couldn't find many other items we could do this for, except for the sub-directories we save artefacts to e.g. ` save_dir='model'` in `save_model_config`. However, these are mostly buried in backend-specific submodules so defining global constants would complicate matters I think.

jklaise commented on alibi_detect/utils/saving.py:
```diff
@@ -948,856 +71,384 @@ def load_detector(filepath: Union[str, os.PathLike], **kwargs) -> Data:
-        Dictionary containing the `n_gmm`, `latent_dim` and `recon_features` parameters.
+        Filepath to directory to save 'config.toml' file in.
 
     Returns
```
I just did a `grep` and it seems the return value is not used anywhere, do we need to keep it?

ascillitoe replied:
Removed the return. Thanks!

jklaise commented on alibi_detect/utils/tensorflow/_saving.py:
```diff
@@ -0,0 +1,945 @@
+
+    if isinstance(model, tf.keras.Model) or isinstance(model, tf.keras.Sequential):
+        model.save(model_path, save_format=save_format)
+    else:
+        logger.warning('No `tf.keras.Model` or `tf.keras.Sequential` detected. No model saved.')
```
Is this branch logically possible? Or are we expecting this to be publicly used?

ascillitoe replied:
We hopefully shouldn't get to this branch, since hopefully if there are any issues picking apart the `model` (i.e. if it's a `UAE` etc) this error will be raised in the function above (`save_model_config`):

https://github.com/SeldonIO/alibi-detect/blob/fc83c707d359d4e5009b7cfb84191748a5625bdd/alibi_detect/utils/tensorflow/_saving.py#L74-L75

However, there might be edge cases I haven't thought of where we successfully get to `save_model` despite there being an issue. This would be fairly problematic wrt to saving though so I have changed from a warning to an error.

jklaise commented on alibi_detect/utils/tensorflow/_saving.py:
```diff
@@ -0,0 +1,945 @@
+    # save classification model
+    model_path = model_path.joinpath('model.h5') if save_format == 'h5' else model_path
+
+    if isinstance(model, tf.keras.Model) or isinstance(model, tf.keras.Sequential):
```
Might be wrong, but iirc `Sequential` is a subtype of `Model`.

ascillitoe replied:
Good spot, I've removed `tf.keras.Sequential` in all places except a few legacy functions.

mauicv commented on alibi_detect/utils/loading.py:
```diff
@@ -0,0 +1,625 @@
+        try:
+            dic = dic[key]
+        except (TypeError, KeyError):
+            return None
```
Is there any risk in ignoring this error?

ascillitoe replied:
Sorry, can you clarify what you mean please?

mauicv replied:
Ah, ignore me! I didn't check properly how you were using it.

jklaise commented on alibi_detect/utils/saving.py:
```diff
@@ -948,856 +71,384 @@ def load_detector(filepath: Union[str, os.PathLike], **kwargs) -> Data:
+        model_cfg, _ = _save_model_config(model, base_path=filepath, input_shape=cfg['input_shape'], backend=backend)
+        cfg['model'] = model_cfg
+
+    # Serialize dataset
+    dataset = cfg.get('dataset', None)
+    if dataset is not None:
+        dataset_cfg, dataset_kwargs = _serialize_function(dataset, filepath, Path('dataset'))
+        cfg.update({'dataset': dataset_cfg})
+        if len(dataset_kwargs) != 0:
+            cfg['dataset']['kwargs'] = dataset_kwargs
```
I don't think I follow! What is a `dataset` and how is it a function?

ascillitoe replied:
Its a callable which is passed as a `kwarg` to some detectors e.g. [ClassiferDrift](https://docs.seldon.io/projects/alibi-detect/en/latest/cd/methods/classifierdrift.html). Usually, it would be a [TFDataset](https://docs.seldon.io/projects/alibi-detect/en/latest/api/alibi_detect.utils.tensorflow.data.html?highlight=TFdataset#alibi_detect.utils.tensorflow.data.TFDataset) or the pytorch equivalent. Since it is pre-registered, `_serialize_function` will save it to the `config.toml` as a function registry (i.e. an `'@...'` string). 

We should probably open an issue to document `TFDataset` etc better, as it could currently be confused with this page:
https://docs.seldon.io/projects/alibi-detect/en/latest/datasets/overview.html

ascillitoe replied:
Correction: `dataset` is not a model. `TFDataset` is a [tf.keras.utils.Sequence](https://www.tensorflow.org/api_docs/python/tf/keras/utils/Sequence), which can be pickled. Hence we are fine here.

mauicv commented on alibi_detect/utils/loading.py:
```diff
@@ -0,0 +1,625 @@
+    filepath = Path(filepath)
+
+    cfg = dict(toml.load(filepath))  # toml.load types return as MutableMapping, force to dict
+    logger.info('Loaded config file from %s', str(filepath))
```
can use f-string

ascillitoe replied:
We don't use f-strings for logging (due to a performance overhead I can't remember the details of). I've changed it to a `.format()` string to be consistent with the other logging calls.

jklaise commented on alibi_detect/utils/saving.py:
```diff
@@ -948,856 +71,384 @@ def load_detector(filepath: Union[str, os.PathLike], **kwargs) -> Data:
+            cfg_token = _save_tokenizer(v, filepath, local_path)
+            kwargs.update({k: cfg_token})
+
+        # Arbitrary function
+        elif callable(v):
+            with open(filepath.joinpath(k + '.dill'), 'wb') as f:
+                dill.dump(v, f)
+            kwargs.update({k: local_path.joinpath(k + '.dill')})
```
Why not use `_serialize_function` here?

jklaise commented on alibi_detect/utils/tests/test_saving.py:
```diff
@@ -1,242 +1,888 @@
+
+
+@fixture
+def preprocess_hiddenoutput(classifier, current_cases, backend):
```
`current_cases` not used in the function body

jklaise commented on alibi_detect/utils/tests/test_saving.py:
```diff
@@ -1,242 +1,888 @@
+                 preprocess_fn=preprocess_fn,
+                 preprocess_at_init=True,
+                 )
+    save_detector(cd, tmp_path)
```
Nice! Didn't know `tmp_path` was available in `pytest`.

ascillitoe replied:
😄

jklaise commented on alibi_detect/utils/tests/test_saving.py:
```diff
@@ -1,242 +1,888 @@
+    Test KSDrift on continuous datasets, with UAE and classifier softmax output as preprocess_fn's. Only this
+    detector is tested with preprocessing strategies, as other detectors should see the same preprocess_fn output.
+
+    Detector is saved and then loaded, with assertions checking that the reinstantiated detector is equivalent.
```
This is definitely out of scope, but I wonder if it's useful/relevant at all to implement the Python dunder methods to test equality of specific detectors? Might be helpful in general for a user to distinguish quickly what is different between two detectors of the same type (essentially a diff between configs).

ascillitoe replied:
This would be very nice. The current equivalence assertions are a real headache. I've added it to the issue list.

mauicv commented on alibi_detect/utils/loading.py:
```diff
@@ -0,0 +1,625 @@
+            val = val.split('.')
+#            val[0] = np if val[0] == 'np' else tf if val[0] == 'tf' else torch if val[0] == 'torch' else None
+            # TODO - add above back in once optional deps are handled properly
+            val[0] = np if val[0] == 'np' else tf if val[0] == 'tf' else None
```
something like:
```py
lib, dtype, *_ = val.split('.')
lib = np if lib == 'np' else tf if lib == 'tf' else None
_set_nested_value(cfg, key, getattr(lib, dtype))
```
would be more readable here.

ascillitoe replied:
Thanks, much nicer!

mauicv commented on alibi_detect/utils/saving.py:
```diff
@@ -948,856 +71,384 @@ def load_detector(filepath: Union[str, os.PathLike], **kwargs) -> Data:
+        func = func.func
     else:
-        return None, None
+        kwargs = {}
```
Why do we have the two cases here? partial function and kwargs / function and no kwargs?

ascillitoe replied:
P.s. leaving unresolved in case of following questions.

mauicv replied:
ok, makes sense!

jklaise commented:
My thoughts on the outstanding decisions:

> * [ ]  Currently the top-level `backend` config field is used to declare the `backend` for backend-specific detectors. But it is also used to set the expected backend for all preprocessing models and kernels etc. Is this OK (at least for now), or do we want specific flags in the `model` (or `preprocess_fn`) configs? Part of this decision might depend on whether we ever envisage one library being used for the backend whilst another is used for preprocessing. I can't see this being sensible for PyTorch/Tensorflow, but perhaps for sklearn?

I think we can defer this discussion to another time. One thing to note is that as there are no API changes, in principle a user should be able to mix and match backends for different steps? Except it would break if saving due to the `backend` logic? @ascillitoe @arnaudvl 

> * [ ]   At the moment the following functions are public:
>   
>   * `save_detector`/`load_detector`  - to save/load detector
>   * `write_config`/`read_config` - to write/read `config.toml` from/into config dict
>   * `validate_config` - to validate config dict
>   
>   All other functions in `saving/loading.py` are private, and the `tensorflow/_saving.py` etc is also private. A number of functions exist to save/load artefact configs, e.g. `_load_model_config`. Making these public could be useful for some users, e.g. so the `model` section of a `config.toml` could be loaded in isolation for debugging, or written runtime model to assist with `config.toml` generation. However making these public will hinder future code changes, so I'm inclined to leave them private unitl the config functionality is more stable?

Agreed, leave private, easier to extend public API than shrink it.


> * [x]  The current options for model type (e.g. `'UAE', 'HiddenOutput', 'custom'`) have been carried over from the legacy save/load code. We could do with rethinking what is required here.

Anything else to do here?

> * [ ]  Find an example where `custom_objects` is needed, and investigate how these objects are defined. It might be easier to remove support for this for now. - _See discussion in comment later on_

Seems the only place is for passing into `tf.keras.models.load_model` as `kwargs`. I would keep this scope as narrow as it currently is, just a question of whether this needs documenting? Given that the only place it's used is internal (https://github.com/SeldonIO/alibi-detect/blob/6b5b2c7b288e15039a115753d14720bb867c535a/alibi_detect/utils/fetching.py#L82-L86) then I don't think it even needs to be? Also, `tf.keras.backend` is a module, does this complicate any of the saving/loading logic? On second thought, if it's only used in the `fetching.py` module, do we even need it as part of the config spec? Perhaps we can get rid of it entirely (can always extend the spec if needed)?

> * [ ]  In `resolve_cfg`, we only attempt to resolve fields in the `config.toml` which are listed in `FIELDS_TO_RESOLVE`. This has the advantage of avoiding infinite recursion, and also allows us to easily tell downstream deps (e.g. MLServer etc) what fields could potentially point to artefacts that will need resolving (such a list has been requested before). However, it is messy, and complicates the resolution of more generic containers such as `custom_objects`. If we assume that only a validated config will be passed to `resolve_cfg` (so we are certain of its structure), I'm wondering if we should go back to more generic recursion here?

Not sure if worth the time to consider this now? Agree it would be nicer, but unclear if we can make a more generic implementation given the fairly custom nature of the detectors. The list of potential artifact fields is a pretty big bonus for downstream applications too.

mauicv commented on alibi_detect/utils/tests/test_saving.py:
```diff
@@ -1,242 +1,888 @@
+    assert cd_load._detector.p_val == P_VAL
+    assert isinstance(cd_load._detector.preprocess_fn, Callable)
+    assert cd_load._detector.preprocess_fn.func.__name__ == 'preprocess_drift'
+#    assert cd.predict(X_h0)['data']['p_val'] == cd_load.predict(X_h0)['data']['p_val']  # Not deterministic
```
Why does setting the seed not make this deterministic?

ascillitoe replied:
Good question! I've checked, and setting all seeds (i.e. `np`, `tf` etc) doesn't guarantee determinism even on cpu. This is a known issue with the learned detectors (https://github.com/SeldonIO/alibi-detect/issues/390), but not clear why it occurs for some others. We really need to look into randomness with Tensorflow/PyTorch properly. I've made a note in the OP to open an issue about it.

mauicv commented on doc/source/overview/config_files.md:
```diff
@@ -0,0 +1,752 @@
+fact a nested field. In other words, `preprocess_fn` is an [Artefact dictionary](dictionaries), and `model` is a field 
+within the `preprocess_fn` dictionary. This layout is required to specify more complex artefacts such as 
+`preprocess_fn`, which themselves rely on other artefacts. TOML files are highly flexible when it comes to specifying 
+dictionaries; each section in a TOML file, demarcated by section headers enclosed in [ ] brackets, is actually a Python 
```
May be better to link to TOML docs that explain how to define sections instead of explaining it bellow?

mauicv commented on doc/source/overview/config_files.md:
```diff
@@ -0,0 +1,752 @@
+
+Note that since `resolved=True`, `validate_config` is now expecting `x_ref` to be a Numpy ndarray instead of a string.
+This second level of validation can be useful as it helps detect problems with loaded artefacts before attempting the
+sometimes time-consuming operation of instantiating the detector. 
```
Do we expect the user to use `resolve_cfg` to obtain a resolved config here? If so should we document that?

jklaise commented:
@ascillitoe FYI we had to do a patch release with a change to the public API due to memory issues which might impact some of this PR: https://github.com/SeldonIO/alibi-detect/pull/474/files

ascillitoe commented:
> @ascillitoe FYI we had to do a patch release with a change to the public API due to memory issues which might impact some of this PR: https://github.com/SeldonIO/alibi-detect/pull/474/files

Yep saw that, I will update the schema's etc, thanks!

ascillitoe commented on alibi_detect/utils/loading.py:
```diff
@@ -0,0 +1,625 @@
+    ['initial_diffs']
+]
+
+# Directories to amend before resolving config (fields to prepend config file dir to)
```
Close. It's actually to prepend the `config.toml` filepath to the directories spec'd within that file.

For example, `config.toml` is stored in `filepath='my_detector'`, which we pass to `load_detector`. Within that `config.toml`, we have `preprocess_fn.model.src = 'model/'`. Since we call `load_detector` from somewhere outside of `filepath`, we need to prepend `filepath` to the directories spec'd inside the `config.toml` i.e. `preprocess_fn.model.src` needs to become `my_detector/model/`. 

The difference between `FIELDS_TO_RESOLVE` and `DIR_FIELDS` is that there are some fields which need resolving but are not directory fields themselves e.g. `embedding` is a `EmbeddingConfig` (we can't spec it as `embedding.dill` directly). The actual directory is `embedding.src`, hence why its in `DIR_FIELDS`. 

Some fields, such as `preprocess_fn.src` can be spec'd directly via `.dill`, or spec'd by setting `preprocess_fn` as a `PreprocessConfig`, hence why `preprocess_fn.src` is in both lists.

ascillitoe replied:
p.s. I've added a similar comment to the code.

ascillitoe commented on alibi_detect/utils/loading.py:
```diff
@@ -0,0 +1,625 @@
+    # Backend and detector type
+    detector_name = cfg.get('name')
+    backend = cfg.pop('backend')  # popping so that cfg left as kwargs + `name` when passed to _init_detector
+    if backend.lower() != 'tensorflow':
+        raise NotImplementedError('Loading detectors with PyTorch or sklearn backend is not yet supported.')
```
Leaving for a future issue in line with https://github.com/SeldonIO/alibi-detect/pull/469#discussion_r837547884.

jklaise commented on alibi_detect/utils/_types.py:
```diff
@@ -2,8 +2,75 @@
+        def validate(cls, val: Any, field: ModelField) -> np.ndarray:
+            return _validate(cls, val, field)
+
+    class PotentialNDArray(Generic[T], np.ndarray):
+        """Like NDArray, but validation errors result in None."""
```
What's the intended use case for this?

ascillitoe replied:
It's literally the same as `NDArray`, but catches the `ValueError` and returns `None`. I've left it in there as thought there might be a use case for this in the future. Its not used at all atm though so we could just remove and add in again if needed?

jklaise replied:
Yea I wasn't sure what the use case is for this, looking at the original docs seems it's used for loading files from disk? Perhaps best to remove if we don't use it.

jklaise commented on alibi_detect/utils/_types.py:
```diff
@@ -2,8 +2,75 @@
+def _validate(cls: Type, val: Any, field: ModelField) -> np.ndarray:
+    if field.sub_fields is not None:
+        dtype_field = field.sub_fields[0]
+        return np.array(val, dtype=dtype_field.type_)
+    else:
+        return np.array(val)
```
I'm wondering if we should use `np.asarray` instead mostly because `np.array` creates a new copy, so if `val` is already an array (possible?) then you would get another copy in memory until the garbage collector comes round. `np.asarray` will not create a copy unless necessary. See https://stackoverflow.com/a/52103839 for more info.

ascillitoe replied:
Good point!

jklaise commented on alibi_detect/utils/schemas.py:
```diff
@@ -1,158 +1,450 @@
-SupportedModels = []
+from alibi_detect.cd.tensorflow import UAE as UAE_tf, HiddenOutput as HiddenOutput_tf
+from transformers import PreTrainedTokenizerBase
+SupportedModels_li = []
```
`li` suffix isn't very suggestive of anything... :)

ascillitoe replied:
It was short for `list` but I guess that isn't very clear!

jklaise commented on doc/source/overview/config_files.md:
```diff
@@ -634,18 +375,21 @@ layers = [-1, -2, -3, -4, -5, -6, -7, -8]
 
-### Validating config files
+(validation)=
+## Validating config files
```
This means the whole section is commented out?

ascillitoe replied:
I've just removed the "Advanced usage" header, and promoted "Validating config files", since we didn't have any other level 3 headings under "Advanced usage". Can change in future if we add more "Advanced topics".

jklaise commented on doc/source/overview/config_files.md:
```diff
@@ -566,14 +307,14 @@ objects can be listed with `registry.get_all()`. For example, `registry.find("my
+:------------------------------------------------------------------- --|:----------------------------------------------|------------|---------|
+| {func}`~alibi_detect.cd.tensorflow.preprocess.preprocess_drift`      | `'@cd.[backend].preprocess.preprocess_drift'` | ✔          | ✔       |
+| {class}`~alibi_detect.utils.tensorflow.kernels.GaussianRBF`          | `'@utils.[backend].kernels.GaussianRBF'`      | ✔          | ✔       |
+| {class}`~alibi_detect.utils.tensorflow.data.TFDataset`               | `'@utils.tensorflow.data.TFDataset'`          | ✔          |         |
```
Currently mis-rendered: 
![image](https://user-images.githubusercontent.com/13080878/163171620-d987802b-9fb7-4fd0-8657-5f25323e7e23.png)

ascillitoe replied:
🤦🏻‍♂️

ascillitoe commented:
> @ascillitoe on the submodules for `save_detector`, `load_detector`, I think it's about time we exported these into a higher public API rather than the user having to import from `alibi_detect.utils.saving/loading`? In `alibi` I've just done `alibi.saving`, if we wanted to do the same perhaps a new subpackage `alibi_detect.saving` could be defined that imports the public functions from the deeper modules?

> @ascillitoe also thinking if we should elevate the registry from `alibi_detect.utils.registry` to e.g. `alibi_detect.registry`?

`alibi_detect.utils.validate`, `alibi_detect.utils.registry`, `alibi_detect.utils.saving`, and `alibi_detect.utils.loading` have now all been moved to `alibi_detect.saving`. For backwards compatiblity I've left `save_detector` and `load_detector` wrapper functions in `alibi_detect.utils.saving` with `DeprecationWarning`'s.

ascillitoe commented on alibi_detect/utils/loading.py:
```diff
@@ -0,0 +1,625 @@
+    """
+    # Load toml if needed
+    if isinstance(filepath, (str, os.PathLike)):
+        config_file = Path(deepcopy(filepath))
```
Good question! Removed.

ascillitoe commented:
> @ascillitoe can I suggest running `isort` at least on the new modules? I know we don't have it in the official guidelines, but perhaps we should...

I've just run this on `alibi_detect.saving`. Otherwise too many changes! Have added note to open an issue too.

ascillitoe commented:
> @ascillitoe a thought I have, if at any point in saving an error is raised then it seems there would be a half-baked saved detector on disk. Should we wrap our top-level saving in `try/except` and do a clean-up if the saving is unsuccessful for any step?

I've added something to delete all new files (not ones already in `filepath`) and delete the `filepath` directory if its empty. Also added a test `test_cleanup`.

ascillitoe commented:
@jklaise @mauicv thanks for all your comments! I've tried to address them all (I've left some unresolved for you to review). There is one oustanding issue @jklaise spotted wrt to saving `dataset`. It currently works, but might not if a user spec'd `dataset` as a custom TensorFlow model (since we currently assume it can be serialized as `dill`).

jklaise commented on alibi_detect/saving/saving.py:
```diff
@@ -49,23 +50,63 @@ def save_detector(detector: Detector, filepath: Union[str, os.PathLike], legacy:
-        save_detector_legacy(detector, filepath)
+    except Exception as error:
+        _cleanup_filepath(orig_files, filepath)
+        print('Saving failed due to the following error. Files/directories have been cleaned up.')
+        raise error
```
Why not put the message in an error, i.e. something like `raise SomeError('text') from error`.

ascillitoe replied:
That'd be much nicer, don't know why I did the above. Will change!

ascillitoe replied:
Replaced with 

```python
raise RuntimeError(f'Saving failed. The save directory {filepath} has been cleaned.') from error
```

Also removed old inaccurate comments regarding the cleanup logic.

jklaise reviewed (approved):
LGTM, but can you remind me if changes to imports using `alibi.saving/loading` as opposed to `alibi.utils.saving/loading` were already made in the example notebooks?

ascillitoe commented:
> LGTM, but can you remind me if changes to imports using `alibi.saving/loading` as opposed to `alibi.utils.saving/loading` were already made in the example notebooks?

Thanks, I had missed those. I'll do a grep and make sure all instances are updated.

ascillitoe commented:
@jklaise I've just resolved the outstanding issue wrt to saving `dataset`'s. Did you want to look at the remaining comments that are still "unresolved", or are they resolved to your satisfaction? 

@arnaudvl do you want to give this a pass now? Or wait until the final merge of the feature branch into `master`?
